### ISSUE TITLE
Check if tests pass before submitting

### ISSUE THREAD

**Issue by alebaffa** [exercism/cli#230]
I just saw a submission empty (I can give you the link if you want). I don't know why the user has submitted an empty solution, but it is now in my list of feedback to give and it makes no sense to leave a comment on it. I think this can be an improvement for the cli.

The cli should check whether the tests pass or not before pushing the code to exercism. If the tests don't pass, it fails displaying a console message like _"sorry, you have to pass all the tests before submitting your solution"_.

What do you think?.

**Comment by kytrinyx** [comment]
Since we have over 20 different languages, it's going to be hard to maintain the functionality that knows how to verify that the tests are passing.

It seems like if someone submits an empty solution, they might need help. I think it would be hugely valuable to add a comment in this situation asking if they got stuck, if there's something that's confusing, and is there any way in which you can help them get started.

**Comment by alebaffa** [comment]
I agree. OK, I will do like that.

**Comment by kytrinyx** [comment]
Cool, I'll go ahead and close this. Thanks for asking!

**Comment by askkaz** [comment]
As I browse other's solutions - it would be nice to have a little flag that lets me know if the solution at least passed the base test cases on the user's system before submission.  This functionality would help people who are trying to learn not waste time looking at bad solutions.  It would also give people trying to help others an easier way to find people in need of help.

**Comment by kytrinyx** [comment]
I agree that this would be helpful, but I don't know how we would do that since running the tests is something that they do completely independent of the exercism CLI.

**Comment by kotp** [comment]
It is something they do independent of the CLI, however, it is received and viewable by the system... I think there is already a bot that gives some advice when you submit, (Rikki??) checks some things already.  Checking for an empty file may be something to look at, specifically in this case.

But, applying the test to the required file name for that exercise may not be out of bounds for Rikki as well?

**Comment by kytrinyx** [comment]
Rikki is not running in a sandboxed environment, and I don't know of any service that provides sandboxed environments for 29 languages.

**Comment by kytrinyx** [comment]
We could look for an empty file--that's low-effort.

**Comment by OJFord** [comment]
Couldn't each language directory have a configuration file that specifies a command to run tests and exit with an appropriate code, so that the CLI just executes that on the exercise directory, and submits only if it exits 0?

For example, Haskell would be `cd $EXERCISE && stack test`; Python would be only slightly more complex - `cd $EXERCISE && python ${EXERCISE//-/_}.py`.

**Comment by kotp** [comment]
I think having it force a no submission due to a failure of the test is counter to what is wanted.  If someone needs help, they should be able to submit their code as it is and request the help.

**Comment by kytrinyx** [comment]
If we were ever to do this, I think we should submit it along with the output of the failure, rather than fail to submit.

**Comment by OJFord** [comment]
> If someone needs help, they should be able to submit their code as it is and request the help.

Does that happen though?

> I think we should submit it along with the output of the failure, rather than fail to submit.

Sure - and I think that could be used to encourage more engagement by still restricting `fetch` - so you could `submit` a failing exercise, and hopefully interact with people to learn what went wrong (or worst case just copy someone else's and resubmit) before progressing to the next one.

Could have a list of "exercises you've completed but others need help with" too, to encourage engagement from the 'other side' - people that can provide help.

**Comment by kytrinyx** [comment]
> Does that happen though?

Yeah, it does.

> Could have a list of "exercises you've completed but others need help with" too, to encourage engagement from the 'other side' - people that can provide help.

This is a good idea, and I believe this is in the works: a filter on the activity stream for "help requested" (which is a flag on the exercise).

**Comment by sshine** [comment]
As I was just about to make a very similar feature request, I'll contribute to this discussion:

We do want people to be able to submit code that fails tests or does not compile. But we also want to encourage that they should try to build and test their solution before they submit in the average case.

Since the build systems are different for every language, perhaps the `exercism` CLI could have an interface that informs it, upon `exercism submit ...`, whether tests were attempted, and whether they succeeded.

Build systems can then gradually support this interface, so that individual language tracks may display warnings like:

> Hey, it looks like you didn't run any tests!
> You should probably do that before you submit.
> This message won't appear again for this exercise.
> You can re-run `exercism submit ...` to submit.

Or:

> Hey, it looks like you didn't pass all your tests!
> This may cause your feedback to mainly be about
> fixing the bugs rather than improving on the solution.
> This message won't appear again for this exercise.
> You can re-run `exercism submit ...` to submit.

The only thing I haven't figured out is how to disable this feature by default, and enable it for tracks for which the build/test systems have been adapted. That is, if we determine this by the presence of meta data in the directory, and with the `exercism` CLI having no prior knowledge of which tracks support what. Since if no build/test command is run, then it would be equivalent to using a track for which the build/test command does not interface with the `exercism` CLI in this way.

Perhaps if this nut can be cracked, we can roll out a useful feature gradually.

**Comment by kytrinyx** [comment]
@sshine these are good thoughts. I still have no idea how to approach it, but I'm going to reopen this so that we can re-think this through (eventually).

**Comment by dogweather** [comment]
I'm 👍 on this and surfacing the info in the web app because of the mentor's UX: Currently, there's no indication whether the code actually works or not. And so from the mentor's perspective, they can't just help with coding ideas, or style hints. They also need to deduce whether the code functions correctly. Or have I been doing it wrong? :-) This is only easy to do on the simpler challenges.

Story
------
As a mentor, I'd like to see a passing/not-passing indicator so that I can give better feedback in less time.

**Comment by dogweather** [comment]
**Implementation idea:** We can check test suite results by checking the Unix status code. I believe that all tests suites return `0` on success. We could have language-specific adapters to run the suites from within the CLI.

**Comment by NobbZ** [comment]
> We could have language-specific adapters to run the suites from within the CLI.

Who writes them? Not for every track there is someone available proficient enough in go to write the adapter.

Or what should this adapter do?

**Comment by dogweather** [comment]
@NobbZ Brainstorming: The CLI fails gracefully. It looks for an adapter implementation for the current language. If it finds that it has none, it simply performs the current behavior. And then the mentor view will show a flag, **Test Suite status: unknown**. "Unknown" can link to a pitch for the mentor to contribute an adapter. :-) So they'll be incentivized.

An adapter, then, would be a Go function which implements a simple interface. In pseudo-code:

```
# @return [true] if entire test suite passes. Otherwise, [false]
run_test_suite() -> boolean
```

**Comment by NobbZ** [comment]
To be honest, I consider the **if** a rather useless information. At least if it fails. In that case as a mentor I'd like to know what cases failed and how. Do we need to attach full test output? Just the failing pieces? Might there be test libraries that split their output between stdin and stdout?

Also starting an external program is not trivial in go, especially not if you want the exit code. It might carry some meaning above failing or not. 1 might be test failures, while 2 means didn't compile at all.

Next thing is, you need to bundle a new release for the CLI whenever the way for running the tests changes.

**Comment by dogweather** [comment]
**Use Case 1:** Many students aren't running any tests:

![screen shot 2019-03-07 at 10 49 32 am](https://user-images.githubusercontent.com/150670/53981248-03ec5400-40c7-11e9-8d59-7c3295b9266f.png)

And in a short-sighted utilitarian sense, why should they? They can submit w/out running them, and then a mentor will give personal feedback. We're not incentivizing them to do so.

I can easily see that this code fails. But I can't do this for more advanced exercises. (And I think I shouldn't have to!)

**Use Case 2:** Many students submit solutions with ad-hoc `main` functions in the code that partially duplicate the test suite. I infer that they're not running the test suite and instead are falling back to bad habits.

**Comment by dogweather** [comment]
Another implementation option is to run the tests completely server-side, like many online learning tools. And we display the result on the solution page.

**Comment by NobbZ** [comment]
Running server side might be very hard. Some (like me) try to stay up to date and use recent compiler versions and language features, while others might be forced to old and even ancient version of a compiler which does not support newer features and might be otherwise incompatible (removed functions are only a small but obvious example).

To be able to use the compiler features I like, is what I like on exercism. Beeing able to have everything on my local environment which I carefully set up to my needs.

**Comment by dogweather** [comment]
@NobbZ  How do you handle these situations as a mentor? I added an example, above, for the second case.

**Comment by kytrinyx** [comment]
Since this is a product question rather than an implementation question, I'm going to go ahead and move this to http://github.com/exercism/exercism/issues
